src/lib/blockchain.coffee

```diff
@@ -246,8 +246,104 @@ module.exports = (dnschain) ->
+                    nsCNAME2IP   = new ResolverStream
+                        name        : 'nsCNAME2IP' # +'-'+(instanceNum++)
+                        stackedDelay: 100
+                    stackedQuery = new ResolverStream
+                        name        : 'stackedQuery' #+'-'+(instanceNum-1)
+                        stackedDelay: 1000
+                        reqMaker    : (nsIP) =>
+                            dns2.Request
+                                question: q
+                                server: address: nsIP
+
+                    nsIPs = gES.merge(gES.readArray(nsIPs), gES.readArray(nsCNAMEs).pipe(nsCNAME2IP))
+
+                    stopRequests = (code) =>
+                        if code
+                            @log.warn gLineInfo("errors on all NS!"), {q:q, code:RCODE_NAME[code]}
+                        else
+                            @log.debug gLineInfo('ending async requests'), {q:q}
+                        rs.cancelRequests(true) for rs in [nsCNAME2IP, stackedQuery]
+                        cb code
+
+                    nsIPs.on 'data', (nsIP) =>
+                        if _.find(gConf.localhosts, (ip)->S(nsIP).startsWith ip)
+                            # avoid the possible infinite-loop on some (perhaps poorly) configured systems
+                            @log.warn gLineInfo('dropping query, NMC NS ~= localhost!'), {q:q, nsIP:nsIP, info:info}
+                        else
+                            stackedQuery.write(nsIP)
+
+                    nsCNAME2IP.on 'failed', (err) =>
+                        @log.warn gLineInfo('nsCNAME2IP error'), {error:err?.message, q:q}
+                        if nsCNAME2IP.errCount == info.ns.length
+                            stopRequests err.code ? NAME_RCODE.NOTFOUND
+
+                    stackedQuery.on 'failed', (err) =>
+                        @log.warn gLineInfo('stackedQuery error'), {error:err?.message, q:q}
+                        if stackedQuery.errCount == info.ns.length
+                            stopRequests err.code ? NAME_RCODE.SERVFAIL
+
+                    stackedQuery.on 'answers', (answers) =>
+                        @log.debug gLineInfo('stackedQuery answers'), {answers:answers}
+                        res.answer.push answers...
+                        stopRequests()
+
+                else if info.ip6
+                    # we have its IP! send reply to client
+                    # TODO: handle more info! send the rest of the
+                    #       stuff in 'info', and all the IPs!
+                    info.ip6 = [info.ip6] if typeof info.ip6 is 'string'
+                    # info.ip.forEach (a)-> res.answer.push gIP2type(q.name, ttl)(a)
+                    res.answer.push (info.ip6.map gIP2type(q.name, ttl, 'AAAA'))...
+                    cb()
+                else
+                    @log.warn gLineInfo('no useful data from nmc_show'), {q:q}
+                    cb NAME_RCODE.NOTFOUND
+            # /end 'AAAA'
```

Comment:
@satindergrewal This is too much code duplication, and that's not OK because it makes maintaining the software way too difficult.

This code is pretty much a copy of the `A` function except for a couple lines changed.

The `AAAA` key should be a function that calls the `A` function, and the `A` function should in turn deduce that this is an IPv6 call based on the variable `q`.